### PR TITLE
Cut Circuit Transform

### PR DESCRIPTION
**Pull request by anthayes92**
**Context:**
This brings together the circuit cutting pipeline allowing a user to add a `@qcut.cut_circuit` decorator to a `qnode` to indicate that a circuit will be cut.

**Description of the Change:**
A batch transforms has been added.

**Benefits:**
A clear and concise user interface

**Possible Drawbacks:**
Circuit cutting is relatively limited to a subset of measurements as of this PR

**Related GitHub Issues:**
Development is [underway](https://github.com/PennyLaneAI/adrs/pull/28) for a batch transform that enables execution of multiple tapes over multiple devices

### CODE REVIEW

File: tests/transforms/test_qcut.py [hunk 1]
```diff
@@ -1799,3 +1803,39 @@ def f(x):
...
+            qml.CNOT(wires=[0, 1])
+            qml.RZ(0.240, wires=0)
+            qml.RZ(0.133, wires=1)
+            return qml.expval(qml.PauliZ(wires=[0]))
+
+        spy = mocker.spy(qcut, "qcut_processing_fn")
+        x = np.array(0.531, requires_grad=True)
+        cut_circuit = qcut.cut_circuit(circuit)
+
+        assert round(cut_circuit(x), 8) == round(float(circuit(x)), 8)
```
Comment (anthayes92) [inline tests/transforms/test_qcut.py:1835]: maybe we should be returning a tensor with `requires_grad` for the cut circuit?

File: tests/transforms/test_qcut.py [hunk 2]
```diff
@@ -1827,10 +1828,12 @@ def circuit(x):
             qml.RZ(0.133, wires=1)
             return qml.expval(qml.PauliZ(wires=[0]))
 
+        spy = mocker.spy(qcut, "qcut_processing_fn")
         x = np.array(0.531, requires_grad=True)
         cut_circuit = qcut.cut_circuit(circuit)
 
         assert round(cut_circuit(x), 8) == round(float(circuit(x)), 8)
+        assert spy.call_count == 1
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:1836]: Nice! There is also: https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once

File: tests/transforms/test_qcut.py [hunk 3]
```diff
@@ -1799,3 +1803,39 @@ def f(x):
...
+        x = np.array(0.531, requires_grad=True)
+        cut_circuit = qcut.cut_circuit(circuit)
+
+        assert round(cut_circuit(x), 8) == round(float(circuit(x)), 8)
+        assert spy.call_count == 1
+
+        gradient = qml.grad(circuit)(x)
+        cut_gradient = qml.grad(cut_circuit)(x)
+
+        assert round(gradient, 8) == round(cut_gradient, 8)
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:1841]: Why not `np.isclose()`?

File: doc/releases/changelog-dev.md [hunk 4]
```diff
@@ -81,6 +81,9 @@
   The postprocessing function for the `cut_circuit` transform has been added.
   [(#2192)](https://github.com/PennyLaneAI/pennylane/pull/2192)
 
+  The `cut_circuit` transform has been added.
+  [(#2216)](https://github.com/PennyLaneAI/pennylane/pull/2216/)
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:85]:
```suggestion
  [(#2216)](https://github.com/PennyLaneAI/pennylane/pull/2216)
```

File: tests/transforms/test_qcut.py (hunk 1 above)
Comment (trbromley) [inline tests/transforms/test_qcut.py:1835]: Also good to have `np.isclose()` here

Comment (trbromley) [reply tests/transforms/test_qcut.py:1835]: Yes need to look into how things are cast. I wouldn't worry for this PR, as long as the grad and functions are identical

File: pennylane/transforms/qcut.py [hunk 5]
```diff
@@ -853,3 +857,85 @@ def qcut_processing_fn(
...
+            -0.506395895364911
+
+        This can be cut using the following transform
+
+        .. code-block:: python
+
+            >>> x = 0.531
+            >>> cut_circuit = qcut.cut_circuit(circuit)
+            >>> cut_circuit(x)
+            0.8623011058543121
```
Comment (trbromley) [inline pennylane/transforms/qcut.py:910]: For console-style codeblocks, we don't need the `.. code-block:: python` decorator.
```suggestion
        >>> x = 0.531
        >>> cut_circuit = qcut.cut_circuit(circuit)
        >>> cut_circuit(x)
        0.8623011058543121
```

File: pennylane/transforms/qcut.py [hunk 6]
```diff
@@ -857,6 +861,86 @@ def qcut_processing_fn(
...
+                qml.CNOT(wires=[0, 1])
+                qml.RZ(0.240, wires=0)
+                qml.RZ(0.133, wires=1)
+                return qml.expval(qml.PauliZ(wires=[0]))
+
+            >>> x = 0.531
+            >>> print(circuit(x))
+            0.8623011058543121
+            >>> print(qml.grad(circuit)(x))
+            -0.506395895364911
```
Comment (trbromley) [inline pennylane/transforms/qcut.py:903]: This console section doesn't need to be in the code block

**Review by trbromley — CHANGES_REQUESTED** [review]
Thanks @anthayes92! Looking good, just a few comments on the docstrings side.

File: doc/releases/changelog-dev.md [hunk 7]
```diff
@@ -81,11 +81,16 @@
   The postprocessing function for the `cut_circuit` transform has been added.
   [(#2192)](https://github.com/PennyLaneAI/pennylane/pull/2192)
 
+
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply doc/releases/changelog-dev.md:84]: Sorry, I was under the impression that `black` handles unnecessary line additions/deletions. Will check manually in future.

File: pennylane/transforms/qcut.py [hunk 8]
```diff
@@ -18,14 +18,16 @@
 
 import copy
 import string
-import warnings
 import uuid
-from typing import Sequence, Tuple, List, Dict, Any, Union, ClassVar
+import warnings
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply pennylane/transforms/qcut.py:22]: yes this should be the most up to date result of applying `isort 5.10.1` (previously using `4.3.4`)

File: pennylane/transforms/qcut.py [hunk 9]
```diff
@@ -34,6 +36,8 @@
 from pennylane.tape import QuantumTape
 from pennylane.wires import Wires
 
+from .batch_transform import batch_transform
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply pennylane/transforms/qcut.py:39]: following the example of [`batch params transform` ](https://pennylane.readthedocs.io/en/stable/_modules/pennylane/transforms/batch_params.html) but happy to apply the suggestion thanks!

File: tests/transforms/test_qcut.py [hunk 10]
```diff
@@ -1632,12 +1634,14 @@ def f(x, y, z):
...
 
         meas_basis = [I, Z, X, Y]
+
         states = [np.outer(s, s.conj()) for s in states_pure]
         zero_proj = states[0]
 
         u1 = qml.RX.compute_matrix(x)
         u2 = qml.RY.compute_matrix(y)
         u3 = qml.RX.compute_matrix(z)
+
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply tests/transforms/test_qcut.py:1644]: this must be my IDE, can slip past pretty easily when it's making changes to files I'm actively working on :face_exhaling:

File: pennylane/transforms/__init__.py [hunk 11]
```diff
@@ -85,6 +85,13 @@
 Transforms for circuit cutting
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+This transform accepts tapes and performs circuit cutting.
+
+.. autosummary::
+    :toctree: api
+
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply pennylane/transforms/__init__.py]: this caused the docs to be displayed incorrectly so reverted these line deletions here: 84988057768820cbd40fd2be1b2a471ea4525abe

File: pennylane/transforms/qcut.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (anthayes92) [reply pennylane/transforms/qcut.py:40]: using `qml.batch_transform` caused a circular import error so added from `pennylane.transforms import batch_transform` here: 84988057768820cbd40fd2be1b2a471ea4525abe

File: pennylane/transforms/qcut.py [hunk 12]
```diff
@@ -857,6 +861,86 @@ def qcut_processing_fn(
     return result
 
 
+@batch_transform
+def cut_circuit(tape: QuantumTape, **kwargs) -> Tuple[Tuple[QuantumTape], Callable]:
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply pennylane/transforms/qcut.py]: added here thanks!  84988057768820cbd40fd2be1b2a471ea4525abe

File: pennylane/transforms/qcut.py [hunk 13]
```diff
@@ -857,6 +861,86 @@ def qcut_processing_fn(
...
+    .. note::
+
+        This function is designed for use as part of the circuit cutting workflow. Check out the
+        :doc:`transforms </code/qml_transforms>` page for more details.
+
+    Args:
+        tape (QuantumTape): The tape of the full circuit to be cut.
+
+    Returns:
+        the tapes corresponding to the circuit fragments as a result of cutting
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply pennylane/transforms/qcut.py]: added here thanks! 84988057768820cbd40fd2be1b2a471ea4525abe

File: pennylane/transforms/qcut.py [hunk 14]
```diff
@@ -857,6 +861,86 @@ def qcut_processing_fn(
...
+        >>> cut_circuit = qcut.cut_circuit(circuit)
+        >>> cut_circuit(x)
+        0.8623011058543121
+
+        Futhermore, the output of the cut circuit is also differentiable:
+
+        .. code-block:: python
+
+            >>> qml.grad(cut_circuit)(x)
+            -0.506395895364911
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply pennylane/transforms/qcut.py]: formatted here thanks 84988057768820cbd40fd2be1b2a471ea4525abe

File: pennylane/transforms/qcut.py (hunk 6 above)
Comment (anthayes92) [reply pennylane/transforms/qcut.py]: removed code block here 84988057768820cbd40fd2be1b2a471ea4525abe

File: pennylane/transforms/__init__.py [hunk 15]
```diff
@@ -85,6 +85,14 @@
 Transforms for circuit cutting
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+This transform accepts QNodes, performs circuit cutting, and returns the result of the original
+uncut circuit
```
Comment (trbromley) [inline pennylane/transforms/__init__.py:89]:
```suggestion
uncut circuit.
```

**Review by trbromley — APPROVED** [review]
Awesome, we have a basic circuit cutter :tada:

**Pull request merged**